compiler/src/dotty/tools/dotc/transform/CacheAliasImplicits.scala

```diff
@@ -46,19 +46,36 @@ class CacheAliasImplicits extends MiniPhase with IdentityDenotTransformer { this
 
   override def phaseName: String = CacheAliasImplicits.name
 
+  private def needsCache(sym: Symbol, rhs: Tree)(using Context): Boolean = rhs.tpe match
+    case rhsTpe @ TermRef(NoPrefix, _)
+    if rhsTpe.isStable => false
+    case rhsTpe @ TermRef(pre: ThisType, _)
+    if rhsTpe.isStable && pre.cls == sym.owner.enclosingClass => false
+    case rhsTpe: ThisType => false
+    case _ => true
+
+  /** Transform
+   *
+   *    given def x = rhs
+   *
+   *  to
+   *
+   *    lazy val x = rhs
+   *
+   *  unless `rhs` has a stable type and is of one of them forms
+   *
+   *    this
+   *    this.y
+   *    y
+   *
+   *  Parameterless given defs are generated during typeclass derivation.
```

Comment:
So this code path is only triggered because of these generated defs? Couldn't they be defined as given lazy vals too?

Reply:
I thought of it but then decided against. Typeclass derivation is very implicit. I felt its better to generate in one way only. This would make a difference if, say, the generated typeclass would have a side effect on instantiation. I have no idea whether something like this would be useful or not. But its easier to reason about if parameterized instances and parameterless instances behave the same.